### PR TITLE
[FP8] Move some logic from layers/vllm/fp8 to layers/common/fp8

### PR DESCRIPTION
**Pull request by lk-chen**
# Description

Similar to #1505 , the common part will be shared by layers/jax and layers/vllm

benchmark with
```
# serve
CURRENT_VLLM_MODEL=Qwen/Qwen3-4B-Instruct-2507-FP8\
MODEL_IMPL_TYPE=vllm vllm serve $CURRENT_VLLM_MODEL \
    --max-model-len=8192 \
    --disable-log-requests \
    --no-enable-prefix-caching

# bench
CURRENT_VLLM_MODEL=Qwen/Qwen3-4B-Instruct-2507-FP8\
vllm bench serve \
    --model $CURRENT_VLLM_MODEL \
    --num-prompts 1000 \
    --dataset-name random \
    --random-input-len 1024 \
    --random-output-len 4096

# eval
lm_eval --model local-completions --model_args base_url=http://localhost:8000/v1/completions,model=Qwen/Qwen3-4B-Instruct-2507-FP8 --tasks gsm8k_cot --batch_size auto --apply_chat_template --num_fewshot 8
```

before
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Failed requests:                         0         
Benchmark duration (s):                  1632.14   
Total input tokens:                      1024000   
Total generated tokens:                  4096000   
Request throughput (req/s):              0.61      
Output token throughput (tok/s):         2509.59   
Peak output token throughput (tok/s):    3697.00   
Peak concurrent requests:                1000.00   
Total token throughput (tok/s):          3136.99   
---------------Time to First Token----------------
Mean TTFT (ms):                          597826.12 
Median TTFT (ms):                        515541.25 
P99 TTFT (ms):                           1362466.45
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          85.57     
Median TPOT (ms):                        77.11     
P99 TPOT (ms):                           121.66    
---------------Inter-token Latency----------------
Mean ITL (ms):                           85.57     
Median ITL (ms):                         65.15     
P99 ITL (ms):                            127.09    
==================================================
```
|  Tasks  |Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|---------|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k_cot|      3|flexible-extract|     8|exact_match|↑  |0.7657|±  |0.0117|
|         |       |strict-match    |     8|exact_match|↑  |0.0440|±  |0.0056|

after
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Failed requests:                         0         
Benchmark duration (s):                  1629.63   
Total input tokens:                      1024000   
Total generated tokens:                  4096000   
Request throughput (req/s):              0.61      
Output token throughput (tok/s):         2513.46   
Peak output token throughput (tok/s):    3642.00   
Peak concurrent requests:                1000.00   
Total token throughput (tok/s):          3141.82                                   <------------- + 0.15%
---------------Time to First Token----------------
Mean TTFT (ms):                          597768.85 
Median TTFT (ms):                        515378.84 
P99 TTFT (ms):                           1361143.03
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          85.42     
Median TPOT (ms):                        77.16     
P99 TPOT (ms):                           121.63    
---------------Inter-token Latency----------------
Mean ITL (ms):                           85.42     
Median ITL (ms):                         65.10     
P99 ITL (ms):                            127.08    
==================================================
```
|  Tasks  |Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|---------|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k_cot|      3|flexible-extract|     8|exact_match|↑  |0.7657|±  |0.0117|
|         |       |strict-match    |     8|exact_match|↑  |0.0440|±  |0.0056|

# Tests

CI

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/common/quantization/fp8.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                                        mesh=self.linear_config.mesh)
+
+        if bias is not None and not layer.skip_bias_add:
+            outs += bias
+        outs = slice_sharded_tensor_for_concatenation(
+            outs, self.linear_config.output_sizes, self.linear_config.n_shards)
+        return jnp.concatenate(outs, axis=-1)
+
+    def _apply_split(self,
+                     layer: torch.nn.Module,
```
Comment (kyuyeunk) [inline tpu_inference/layers/common/quantization/fp8.py]: don't think it's a good idea to include `torch` dependency in the `layer/common` directory.

File: tpu_inference/layers/common/linear.py [hunk 2]
```diff
@@ -45,7 +49,9 @@ def sharded_quantized_matmul(x: jax.Array, w_q: jax.Array, w_s: jax.Array,
     scale_sharding = P(out_axis, )
     out_sharding = P(None, out_axis)
 
-    x = jax.lax.with_sharding_constraint(x, NamedSharding(mesh, x_sharding))
+    x = jax.lax.with_sharding_constraint(
```
Comment (kyuyeunk) [inline tpu_inference/layers/common/linear.py:52]: does flax path run with [jax.set_mesh](https://docs.jax.dev/en/latest/_autosummary/jax.set_mesh.html)? I would assume that's why it doesn't need mesh when running `jax.lax.with_sharding_constraint`?

Note that i'm also considering to move torchax path into that paradigm

**Comment by jrplatin** [comment]
LGTM -- you might also want to check that the TorchAX path doesn't show any regression

**Review by jrplatin — APPROVED** [review]

File: tpu_inference/layers/common/quantization/fp8.py (hunk 1 above)
Comment (lk-chen) [reply tpu_inference/layers/common/quantization/fp8.py]: removed, using bias==None to determine if bias is needed.

File: tpu_inference/layers/common/linear.py (hunk 2 above)
Comment (lk-chen) [reply tpu_inference/layers/common/linear.py:52]: Yes I plan to use `set_mesh` for jax path.
